k2/csrc/aux_labels.cc

```diff
@@ -6,14 +6,89 @@
 
 #include "k2/csrc/aux_labels.h"
 
+#include <algorithm>
 #include <numeric>
+#include <utility>
 #include <vector>
 
 #include "glog/logging.h"
 #include "k2/csrc/fsa.h"
+#include "k2/csrc/fsa_util.h"
+#include "k2/csrc/properties.h"
+
+namespace {
+
+/*
+  This function counts how many extra states we need to create for each state in
+  the input FSA when we invert an FSA. Generally, if an entering arc of state
+  `i` in the input FSA has n olabels, then we need to create n-1 extra states
+  for state `i`.
+
+  @param [in] fsa_in      Input FSA
+  @param [in] labels_in   Aux-label sequences for the input FSA
+  @param [out] expanded_states  For state `i` in `fsa_in`, we need to create
+                                extra `expaned_states[i]` states in the output
+                                inverted FSA.
+*/
+static void CountExpandedStatesNum(const k2::Fsa &fsa_in,
+                                   const k2::AuxLabels &labels_in,
+                                   std::vector<int32_t> *expanded_states) {
+  CHECK_EQ(expanded_states->size(), fsa_in.NumStates());
+  auto &states = *expanded_states;
+  for (int32_t i = 0; i != fsa_in.arcs.size(); ++i) {
+    const auto &arc = fsa_in.arcs[i];
+    int32_t pos_start = labels_in.start_pos[i];
+    int32_t pos_end = labels_in.start_pos[i + 1];
+    states[arc.dest_state] += std::max(0, pos_end - pos_start - 1);
+  }
+}
+
+/*
+  Map the state in the input FSA to state in the output inverted FSA.
+
+  @param [in] expanded_states  Output of function `CountExpandedStatesNum`
```

Comment:
The name num_extra_states would be clearer than `expanded_states`.
Also I think CountExtraStates() would be  better for that other function.

Reply:
Sure